video.c

```diff
@@ -1,1185 +1,1226 @@
+		// additional bytes to reach the correct line of the tile
+		uint32_t y_add = (yy * props->tilew * props->bits_per_pixel) >> 3;
+		// additional bytes to reach the correct column of the tile
+		uint16_t x_add = (xx * props->bits_per_pixel) >> 3;
+		uint32_t tile_offset = tile_start + y_add + x_add;
+		uint8_t s = video_space_read(props->tile_base + tile_offset);
+
+		// Apply Palette Offset
+		if (palette_offset) {
+			uint8_t col_index = convert_tile_byte_to_indexed_color(props, xx, s);
+			if (col_index > 0 && col_index < 16) col_index += palette_offset << 4; // Note: Should this be an assert?
+			layer_line[layer][x] = col_index;
+		} else {
+			layer_line[layer][x] = convert_tile_byte_to_indexed_color(props, xx, s);
+		}
+	}
+
+}*/
+
+static inline uint8_t calculate_line_col_index(uint8_t spr_zindex, uint8_t spr_col_index, uint8_t l1_col_index, uint8_t l2_col_index)
+{
+	uint8_t col_index = 0;
+	switch (spr_zindex) {
+		case 3:
+			col_index = spr_col_index ?: l2_col_index ?: l1_col_index;
+			break;
+		case 2:
+			col_index = l2_col_index ?: spr_col_index ?: l1_col_index;
+			break;
+		case 1:
+			col_index = l2_col_index ?: l1_col_index ?: spr_col_index;
+			break;
+		case 0:
+			col_index = l2_col_index ?: l1_col_index;
+			break;
+	}
+	return col_index;
+}
+
+static void
+render_line(uint16_t y)
+{
+	uint8_t out_mode = reg_composer[0] & 3;
+
+	uint8_t border_color = reg_composer[3];
+	uint16_t hstart = reg_composer[4] << 2;
+	uint16_t hstop = reg_composer[5] << 2;
+	uint16_t vstart = reg_composer[6] << 1;
+	uint16_t vstop = reg_composer[7] << 1;
+
+	int eff_y = (reg_composer[2] * (y - vstart)) / 128;
+
+	bool sprite_line_empty = (!(reg_composer[0] & 0x40));
+	if (!sprite_line_empty) render_sprite_line(eff_y);
+
+	bool layer_line_empty[2];
+	layer_line_empty[0] = (reg_composer[0] & (0 ? 0x20 : 0x10)) == 0;
+	layer_line_empty[1] = (reg_composer[0] & (1 ? 0x20 : 0x10)) == 0;
+
+	if (!layer_line_empty[0]) {
+		struct video_layer_properties *props = &layer_properties[0];
+		if (props->bitmap_mode) render_layer_line_bitmap(0, eff_y, props);
+		else if (props->text_mode) render_layer_line_text_mode(0, eff_y, props);
+		else render_layer_line(0, eff_y, props);
+	}
+	if (!layer_line_empty[1]) {
+		struct video_layer_properties *props = &layer_properties[1];
+		if (props->bitmap_mode) render_layer_line_bitmap(1, eff_y, props);
+		else if (props->text_mode) render_layer_line_text_mode(1, eff_y, props);
+		else render_layer_line(1, eff_y, props);
+	}
+
+	uint8_t col_line[SCREEN_WIDTH];
+
+	if (video_palette.dirty) {
+		refresh_palette();
+	}
+
+	// If video output is enabled, calculate color indices for line.
+	if (out_mode != 0) {
+		uint8_t spr_col_index = 0;
+		uint8_t l1_col_index = 0;
+		uint8_t l2_col_index = 0;
+		uint8_t spr_zindex = 0;
+
+		// Calculate color without border.
+		for (uint16_t x = 0; x < SCREEN_WIDTH; x += 1) {
+			const int eff_x = (reg_composer[1] * (x - hstart)) / 128; // What is this 128, should this be SCREEN_WIDTH?

```

Comment:
The 128 here is the horizontal scaling factor.  A reg_composer[1] of 128 equals 1 input pixel to 1 output pixel scaling, 64 equals 1:2, and so on.